ssh/tailssh/tailssh.go

```diff
@@ -259,172 +255,197 @@ func (c *conn) vlogf(format string, args ...any) {
 }
 
-// doPolicyAuth verifies that conn can proceed with the specified (optional)
-// pubKey. It returns nil if the matching policy action is Accept or
-// HoldAndDelegate. If pubKey is nil, there was no policy match but there is a
-// policy that might match a public key it returns errPubKeyRequired. Otherwise,
-// it returns errDenied.
-func (c *conn) doPolicyAuth(ctx ssh.Context, pubKey ssh.PublicKey) error {
-	if err := c.setInfo(ctx); err != nil {
-		c.logf("failed to get conninfo: %v", err)
-		return errDenied
-	}
-	a, localUser, err := c.evaluatePolicy(pubKey)
-	if err != nil {
-		if pubKey == nil && c.havePubKeyPolicy() {
-			return errPubKeyRequired
-		}
-		return fmt.Errorf("%w: %v", errDenied, err)
-	}
-	c.action0 = a
-	c.currentAction = a
-	c.pubKey = pubKey
-	if a.Message != "" {
-		if err := ctx.SendAuthBanner(a.Message); err != nil {
-			return fmt.Errorf("SendBanner: %w", err)
+// welcomeBanner looks for a welcome banner to display prior to authentication.
+// For example, if SSH session recording is enabled, control will give us an
+// action message informing the user of this.
+// This function actually begins authentication in order to make sure it knows
+// if there's a banner to send.
+func (c *conn) welcomeBanner(cm gossh.ConnMetadata) (banner string) {
+	defer func() {
+		if c.action0 != nil {
+			banner = c.action0.Message
+		} else {
+			banner = ""
 		}
+	}()
+
+	if c.insecureSkipTailscaleAuth {
+		return
 	}
-	if a.Accept || a.HoldAndDelegate != "" {
-		if a.Accept {
-			c.finalAction = a
-		}
-		lu, err := userLookup(localUser)
-		if err != nil {
-			c.logf("failed to look up %v: %v", localUser, err)
-			ctx.SendAuthBanner(fmt.Sprintf("failed to look up %v\r\n", localUser))
-			return err
-		}
-		gids, err := lu.GroupIds()
-		if err != nil {
-			c.logf("failed to look up local user's group IDs: %v", err)
-			return err
-		}
-		c.userGroupIDs = gids
-		c.localUser = lu
-		return nil
+
+	if err := c.setInfo(cm); err != nil {
+		c.logf("failed to get conninfo: %v", err)
+		c.action0Error = errDenied("")
+		return
 	}
-	if a.Reject {
-		c.finalAction = a
-		return errDenied
+
+	c.action0, c.action0Error = c.authenticate()
+	if c.action0Error == nil {
+		if c.action0.Accept || c.action0.Reject {
+			c.finalAction = c.action0
+		}
 	}
-	// Shouldn't get here, but:
-	return errDenied
+
+	return
 }
 
 // ServerConfig implements ssh.ServerConfigCallback.
 func (c *conn) ServerConfig(ctx ssh.Context) *gossh.ServerConfig {
-	return &gossh.ServerConfig{
-		NoClientAuth:           true, // required for the NoClientAuthCallback to run
-		NextAuthMethodCallback: c.nextAuthMethodCallback,
+	cfg := &gossh.ServerConfig{
+		BannerCallback:       c.welcomeBanner,
+		NoClientAuth:         true, // required for the NoClientAuthCallback to run
+		NoClientAuthCallback: c.noClientAuthCallback,
+		PublicKeyCallback:    c.publicKeyCallback,
+	}
+	if c.anyPasswordIsOkay {
+		cfg.PasswordCallback = c.passwordCallback
 	}
```

Comment:
I believe `ServerConfig` is called before `noClientAuthCallback`, which sets `anyPasswordIsOkay`. So this condition here will never be `true`.
It should be safe to just remove it here, since we'll do the `PartialSuccessError` in `noClientAuthCallback` as needed.